builder: add simple make jobserver

rhubert opened the pull request:
The make jobserver mechanism need two pipes. As long as it can read from
the request pipe new jobs are started. Once a job has finished a byte is
written back into the response pipe.

fixes: #257

rhubert commented:
Well, this basically works... But I'm unsure if the `__runners` semaphore should also be converted to a read on the pipe?

rhubert commented:
>  Now the tricky part is where the pipe is empty. In this case the task must be put to sleep and the pipe must be watched with [loop.add_reader()](https://docs.python.org/3.5/library/asyncio-eventloop.html#asyncio.AbstractEventLoop.add_reader) to get a callback when the pipe is readable again.

In this case the `read` is already blocking? I pushed some changes to make use of the Jobserver for bob tasks.

> Even make will only forward these if it understands that the callee understands the job protocol (see Job Slots).

Are you referring to `First, only command lines that make understands to be recursive invocations of make..` ? I have not fully understood what the author wanted to say with this paragraph. But a simple Makefile
```
all: foo.sh
	./foo.sh
```
and foo.sh as
``` 
#!/bin/sh
printenv MAKEFLAGS
```
produces:
```bash
$ make -j 8
./foo.sh
 -j8 --jobserver-auth=3,4
```

> So we need something like a jobServer recipe property to selectively enable the support. Otherwise we may break existing recipes.

I think this is indeed need. Otherwise existing recipes with `make -j $(nprocs)` produce a warning like: `make: warning: -j8 forced in submake: resetting jobserver mode.`

But I'm unsure if this should be a property or a policy?

jkloetzke commented:
> > Now the tricky part is where the pipe is empty. In this case the task must be put to sleep and the pipe must be watched with [loop.add_reader()](https://docs.python.org/3.5/library/asyncio-eventloop.html#asyncio.AbstractEventLoop.add_reader) to get a callback when the pipe is readable again.
> 
> In this case the `read` is already blocking?

Yes, but it must not block. Bob uses coroutines internally. When Bob blocks on `read()` then all internal processing and dispatching will stop. Instead the currently active coroutine must be suspended so that another coroutine can be scheduled.

> I pushed some changes to make use of the Jobserver for bob tasks.

I'm a bit confused about the thread that shuffles bytes out and into the pipe. I don't think this is necessary.

> > Even make will only forward these if it understands that the callee understands the job protocol (see Job Slots).
> 
> Are you referring to `First, only command lines that make understands to be recursive invocations of make..` ? I have not fully understood what the author wanted to say with this paragraph. But a simple Makefile

Yes, exactly. I was also confused initially but it makes sense. The `MAKEFLAGS` variable is always there. What is passed conditionally are the file descriptors. Without the `+` the file descriptors are closed. You can see that by inspecting `/proc/self/fd` from the shell script script.

I would not do it this way in Bob, though. If job mode is enabled the `MAKEFLAGS` are set and the file descriptors are passed. Otherwise neither is done.

> > So we need something like a jobServer recipe property to selectively enable the support. Otherwise we may break existing recipes.
> 
> I think this is indeed need. Otherwise existing recipes with `make -j $(nprocs)` produce a warning like: `make: warning: -j8 forced in submake: resetting jobserver mode.`
> 
> But I'm unsure if this should be a property or a policy?

It should be a property. It should be selectively enabled by the `make` class which can then refrain from passing `-j`. Some recipes require serial builds for their packages so they need to disable job mode explicitly. A policy would be too broad I guess.

rhubert commented:
OK, I reworked the semaphore to avoid blocking and added a property which now affects checkout, build and package step in the same way. Not sure if this is a good idea :thinking: 

> I'm a bit confused about the thread that shuffles bytes out and into the pipe. I don't think this is necessary.

I'm open for other ideas.... :grinning: 

> I'm not sure what should happen if the recipes define `MAKEFLAGS` or if Bob is called from a recursive make where it's already set.

Append `--jobserver-auth` to the existing `MAKEFLAGS` envirionment? Maybe we can also check if  `jobserver-auth` is already in `MAKEFLAGS` and throw an error in this case? 

If bob is called from a recursive make as a user I would assume the provided fd's are used and passed. But is this really a use case or just something to note in the documentation?

jkloetzke commented on pym/bob/input.py:
```diff
@@ -1559,6 +1567,7 @@ def __init__(self, recipe, tools, diffTools, sandbox, diffSandbox,
         self.states = states
         self.pkgId = pkgId
         self.fingerprintMask = fingerprintMask
+        self.jobServer = useJobServer
```
There is no need to store the value in the package because it will always be the same as the recipe value. Just access `self.recipe...` to get the value.

jkloetzke commented on pym/bob/cmds/build/builder.py:
```diff
@@ -104,6 +108,100 @@ class RestartBuildException(Exception):
 class CancelBuildException(Exception):
     pass
 
+class JobServer:
+    def jobserver_thread(self):
+        while self.__running:
+            os.write(self.__w_server, os.read(self.__r_server, 1))
+
+    def __init__(self, jobs):
+        self.__running = True
+        self.thread = threading.Thread(target=self.jobserver_thread, args =())
+        self.__r_make, self.__w_server = os.pipe()
+        self.__r_server, self.__w_make = os.pipe()
+        os.write(self.__w_server, bytes(jobs))
+        self.thread.start()
+
+    def getMakeFd(self):
+        return [ self.__r_make, self.__w_make ]
+
+    def stop(self):
+        self.__running = False
+        os.write(self.__w_make, bytes(0x30))
+        self.thread.join()
```
Why not just one pipe, load it with the initial bytes and we're good?

jkloetzke commented on pym/bob/cmds/build/builder.py:
```diff
@@ -104,6 +108,100 @@ class RestartBuildException(Exception):
+
+class JobServerSemaphore:
+    def __init__(self, value, fds):
+        if sys.platform == "win32":
```
I would make the distinction in `Builder.cook()` on Windows we create the `asyncio.BoundedSemaphore ` as before, otherwise the new `JobServerSemaphore`. Then this class here does not need to make this distinction and will get smaller.

jkloetzke commented on pym/bob/cmds/build/builder.py:
```diff
@@ -104,6 +108,100 @@ class RestartBuildException(Exception):
+            self.__waitersCnt = 0
+            self.__jobsAquired = 0
+            self.__fds = fds
+            fcntl.fcntl(self.__fds[0], fcntl.F_SETFL,
+                    fcntl.fcntl(self.__fds[0], fcntl.F_GETFL) | os.O_NONBLOCK)
```
`os.set_blocking`

jkloetzke commented on pym/bob/cmds/build/builder.py:
```diff
@@ -104,6 +108,100 @@ class RestartBuildException(Exception):
+                ret = os.read(self.__fds[0], 1)
+            except BlockingIOError:
+                pass
+            if not ret:
```
This will evaluate to `True` if `ret == 0`. Did you mean `ret is None`?

jkloetzke commented on pym/bob/cmds/build/builder.py:
```diff
@@ -33,6 +33,10 @@
 import tempfile
+import threading
+
+if sys.platform is not "win32":
```
While it will technically work should compare by value (`==`) and not by identity.

jkloetzke commented on pym/bob/cmds/build/builder.py:
```diff
@@ -104,6 +108,100 @@ class RestartBuildException(Exception):
+
+    @staticmethod
+    def jobavailableCallback(self):
+        ret = None
+        try:
+            ret = os.read(self.__fds[0], 1)
+        except BlockingIOError:
+            pass
+        if ret:
+            with self.__lock:
+                self.__waitersCnt -= 1
+                if self.__waitersCnt == 0:
+                    asyncio.get_event_loop().remove_reader(self.__fds[0])
+            self.__sem.release()
```
What about this?

```suggestion
        try:
            while __waitersCnt:
                os.read(self.__fds[0], 1)
                self.__waitersCnt -= 1
                self.__sem.release()
        except BlockingIOError:
            pass
        if self.__waitersCnt == 0:
            asyncio.get_event_loop().remove_reader(self.__fds[0])
```

What is still not right is that we have to remember the values that were extracted from the pipe and use them when writing back to the pipe.

rhubert replied:
Haven't tested this but I think it will not work as os.read is not blocking? So when ever a token is back in the pipe all waiters are released, aren't they? At least the return of `os.read` must be checked?

jkloetzke replied:
My assumption was that either we get exactly one byte from the pipe (and hence we got a ticket) or `read` will throw `BlockingIOError` because the pipe was empty and the call would have blocked. If this assumption does not hold then the code above is indeed garbage.

jkloetzke commented:
> OK, I reworked the semaphore to avoid blocking and added a property which now affects checkout, build and package step in the same way. Not sure if this is a good idea thinking

I don't see a problem with this at the moment. We can always add more fine grained options afterwards.

> > I'm not sure what should happen if the recipes define `MAKEFLAGS` or if Bob is called from a recursive make where it's already set.
> 
> Append `--jobserver-auth` to the existing `MAKEFLAGS` envirionment? Maybe we can also check if `jobserver-auth` is already in `MAKEFLAGS` and throw an error in this case?

I would just append it. If the user defines `MAKEFLAGS` with something incompatible then it's his fault.

> If bob is called from a recursive make as a user I would assume the provided fd's are used and passed. But is this really a use case or just something to note in the documentation?

We could just use the fds instead of creating a pipe on our own. This would imply that we somehow have to cope with the implicit token that Bob will get, though.

I've skimmed through the changes. It's no exhaustive review but this is definitely going in the right direction...

rhubert commented on pym/bob/cmds/build/builder.py:
```diff
@@ -126,6 +129,79 @@ def stop(self):
+        else:
+            with self.__lock:
+                if self.__waitersCnt != 0:
+                    self.__waitersCnt = -1;
```
:facepalm: :facepalm: :facepalm: :facepalm: 
it took me several hours to find the reason why bob sometimes didn't finish... 
:angry: :angry: :angry: :angry:

rhubert commented:
Thanks for the review - again a lot of code could be removed :) 

The use of make provided file descriptors is still missing. 
 > Even make will only forward these if it understands that the callee understands the job protocol (see Job Slots). 

Do you know how make makes this decision? If make foes not understand bob understands the job protocol we don't need to implement this? 

I'm also thinking about a test for this but I have no idea how to count the running make jobs. ATM I'm just using a Makefile which starts a few sleep's and monitor top...

jkloetzke commented:
> Thanks for the review - again a lot of code could be removed :)

You're welcome. :)

> The use of make provided file descriptors is still missing.
> 
> > Even make will only forward these if it understands that the callee understands the job protocol (see Job Slots).
> 
> Do you know how make makes this decision? If make foes not understand bob understands the job protocol we don't need to implement this?

AFAICT it consideres invocations of "make" or "$(MAKE)" and every command that is prefixed by a "+". I could imagine someone writes a recursive makefile where Bob is only a part of it. So I think it's worth the effort.

> I'm also thinking about a test for this but I have no idea how to count the running make jobs. ATM I'm just using a Makefile which starts a few sleep's and monitor top...

Hmm, a black box test will be difficult. I would just do a smoke test there. Instead a unit test may be a better approach where the relevant parts can be mocked.

jkloetzke commented on pym/bob/input.py:
```diff
@@ -2212,6 +2225,9 @@ def coDet(r):
             self.__relocatable = self.__recipeSet.getPolicy('allRelocatable') \
                 or not self.__provideTools
 
+        if self.__jobServer is None:
```
I'm sure there was class inheritance code before. Looks like it got lost in the last refactoring. Currently the property will not work when set in a (anonymous) class.

jkloetzke commented on pym/bob/invoker.py:
```diff
@@ -177,6 +178,10 @@ async def __runCommand(self, args, cwd, stdout=None, stderr=None,
             env = _env
 
+        if len(self.__makeFds) == 2:
+            env["MAKEFLAGS"] = "".join([env.get("MAKEFLAGS") if env.get("MAKEFLAGS") is not None else ""]) + \
+                "-j --jobserver-auth=" + ",".join([str(fd) for fd in self.__makeFds])
```
Shouldn't there be a space between multiple options? I also thought that make forwards the number of jobs, e.g. `-j4` instead of an unbounded `-j`.

jkloetzke commented on pym/bob/invoker.py:
```diff
@@ -305,6 +307,12 @@ async def executeStep(self, mode, clean=False, keepSandbox=False):
                 emptyDirectory(self.__spec.workspaceWorkspacePath)
 
+            if len(self.__makeFds) == 2:
+                self.__spec.env["MAKEFLAGS"] = ("".join([self.__spec.env.get("MAKEFLAGS") if
+                        self.__spec.env.get("MAKEFLAGS") is not None else ""])
+                    + " -j" + str(self.__makeJobs) + " --jobserver-auth="
+                    + ",".join([str(fd) for fd in self.__makeFds]))
```
Hmm, I've played a bit around and it looks like `make` will parse the `MAKEFLAGS` that it gets, forwards some of the options it knows and overwrites the rest. Maybe we should do the same?

rhubert replied:
I don't think we should remove or edit the `MAKEFLAGS` defined in the recipe or somewhere else? What sounds reasonable is to overwrite `-j` and `--jobserver-auth` if they are already in the `MAKEFLAGS`?

jkloetzke replied:
I'm not sure yet how to handle it properly. By default I would probably only forward options that are understood by Bob. For example if you define variables on the command line, e.g. `make -j3 FOO=bar`, you will get `MAKEFLAGS= -j3 --jobserver-auth=3,4 -- FOO=bar`. That will influence the build and we should probably strip the assignment out.

OTOH you are probably right that we should not overwrite what a user did specify in the recipe or on the command line.

So maybe we should just use `-j` and `--jobserver-auth` from the OS environment. But if the user specified something in the recipes then we amend `MAKEFLAGS` and keep the rest...

rhubert commented:
I added some code to make use of `MAKEFLAGS` provided to bob. I'm not sure if it's correct to use `-j` from `MAKEFLAGS` instead of `-j` from cmdline in case we have both?

jkloetzke commented:
> I added some code to make use of `MAKEFLAGS` provided to bob. I'm not sure if it's correct to use `-j` from `MAKEFLAGS` instead of `-j` from cmdline in case we have both?

I made a small test and if I pass `-j2` to a sub-make when the top-make is run with `-j3` then I get the following:

```
$ LANG= make -j3 FOO=bar
sh -c 'echo "top: MAKEFLAGS=$MAKEFLAGS"'
top: MAKEFLAGS= -j3 --jobserver-auth=3,4 -- FOO=bar
make -C subdir -j2
make[1]: warning: -j2 forced in submake: resetting jobserver mode.
make[1]: Entering directory '/home/jkloetzke/tmp/subdir'
sh -c 'echo "sub: MAKEFLAGS=$MAKEFLAGS"'
sub: MAKEFLAGS=w -j2 --jobserver-auth=5,6 -- FOO=bar
make[1]: Leaving directory '/home/jkloetzke/tmp/subdir'
```

I think we should do the same, including the warning.

rhubert commented:
:bellhop_bell:

jkloetzke commented on pym/bob/cmds/build/build.py:
```diff
@@ -156,6 +158,20 @@ def _downloadArgument(arg):
+                    makeFds = [int(fds.group(1)), int(fds.group(2))]
+                else:
+                    print("warning: -j" + str(args.jobs) + " forced: resetting jobserver mode.");
+
```
I think we have to check if the file descriptors are actually open. If the `bob` call is not prefixed with `+` in the `Makefile` then `make` will still set `MAKEFLAGS` but the file descriptors are not open. Something along these lines should work (untested):

```python
try:
  if not all(stat.S_ISFIFO(os.stat(f).st_mode) for f in makeFds): makeFds = None
except OSError:
  makeFds = None
```

jkloetzke commented on pym/bob/cmds/build/builder.py:
```diff
@@ -104,6 +104,79 @@ class RestartBuildException(Exception):
+    def getMakeFd(self):
+        return self.__makeFds
+
+class JobServer:
```
Maybe name it `InternalJobServer` for consistency?

```suggestion
class InternalJobServer:
```

jkloetzke commented on test/test_cmd_build_builder_jobserver.py:
```diff
@@ -0,0 +1,34 @@
+        _run(self.__t.testAcquire())
+        self.__t.release()
+        _run(self.__t.testAcquire(2))
+        self.__t.release()
```
Is the release deliberately unbalanced?

jkloetzke reviewed (commented):
Sorry. I didn't see that you've already pushed weeks ago. :see_no_evil: Anyway, just some last minor comments...

rhubert commented:
Don't stress. I was also away a couple of weeks ... :)

I pushed a new commit for the latest issues. If it's ok I'll squash them..

jkloetzke commented:
> Don't stress. I was also away a couple of weeks ... :)

:laughing: 

> I pushed a new commit for the latest issues. If it's ok I'll squash them..

Looks good. :+1:  I'll merge after the squash...

jkloetzke commented:
:tada: Thanks!
